### PR TITLE
A few small external tables tweaks to aid user friendliness

### PR DESCRIPTION
**Pull request by danielgustafsson**
A few small patches I had lying around after reading the external tables code. There are no changes to functionality, just trying to help the user a bit.

**Move duplicate handling of LOCATION to the parsing stage:** Creating an external table with duplicate location URIs is a syntax error so move the duplicate check to the parsing step. Simplifies the external table command code as well as provides an error message with context marker to the user.

**Issue a WARNING for external tables with too many URIs:** External tables with protocols file or http which specify more URIs than segments won't be allowed to query until the cluster has been extended. Rather than silently allow creation, issue a warning and
hint to the user to assist troubleshooting. Writable tables error out for the same problem but avoiding to change behavior this late in the release cycle seems like a good idea.

### CODE REVIEW

**Comment by ivannovick** [comment]
cool

File: src/backend/parser/gram.y [hunk 1]
```diff
@@ -8917,7 +8917,15 @@ DeallocateStmt: DEALLOCATE name
...
 cdb_string_list:
 			cdb_string							{ $$ = list_make1($1); }  
-			| cdb_string_list ',' cdb_string	{ $$ = lappend($1, $3); }
+			| cdb_string_list ',' cdb_string
+				{
+					if (list_member($1, $3))
+						ereport(ERROR,
+								(errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+								 errmsg("duplicate location uri"),
+								 scanner_errposition(@2)));
```
Comment (hlinnaka) [inline src/backend/parser/gram.y]: Should be "scanner_errposition(@3)". Now the error position points to the comma, which is weird.

File: src/backend/commands/exttablecmds.c [hunk 2]
```diff
@@ -359,6 +359,34 @@ DefineExternalRelation(CreateExternalStmt *createExtStmt)
...
+		{
+			Value	*loc = lfirst(list_head(exttypeDesc->location_list));
+			Uri 	*uri = ParseExternalTableUri(loc->val.str);
+
+			if (uri->protocol == URI_FILE || uri->protocol == URI_HTTP)
+			{
+				if (getgpsegmentCount() < list_length(exttypeDesc->location_list))
+					ereport(WARNING,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("Number of locations, %d, exceed number "
```
Comment (hlinnaka) [inline src/backend/commands/exttablecmds.c]: Some typos and stylistic issues here:

* per style guide, error messages should begin with lowercase. (admittedly we violate that in many GPDB error messages)
* wont -> won't
* exceed -> exceeds (not 100% sure about this, but, the "exceeds" refers to "number of locations", which is singluar, rather than "locations", which would be plural).
* "Queries won't be allow", well, queries in general are allowed, just not on this table, maybe rephrase. And this probably belongs in the hint.

Overall, I'd suggest something like:

```
ereport(WARNING,
  (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
   errmsg("number of locations (%) exceeds the number of segments (%d)"),
   errhint("The table cannot be queried until cluster is expanded so that there are at least as many segments as locations.")));
```

File: src/backend/parser/gram.y (hunk 1 above)
Comment (danielgustafsson) [inline src/backend/parser/gram.y]: Of course, not sure how I convinced myself this was right.. =)

File: src/backend/commands/exttablecmds.c (hunk 2 above)
Comment (danielgustafsson) [inline src/backend/commands/exttablecmds.c]: Thanks, that reads much better. Updated in rebase.

**Comment by danielgustafsson** [comment]
Comments addressed and branch rebased on top of master

**Comment by danielgustafsson** [comment]
Pushed to master, thanks for review!
